File: packages/react/test/myers.test.ts
import { describe, expect, it } from 'vitest';
import { diffSequence } from 'diffine-react';

/**
 * The length of the longest common subsequence, worked out the slow way.
 *
 * The engine finds the smallest set of edits, and the smallest set of edits is
 * everything the two sequences do not have in common — so an answer that pairs
 * up fewer tokens than this is a wrong answer, and one that pairs up more is
 * impossible. This is quadratic and reads straight off the definition, which is
 * exactly what a check on a clever implementation should be.
 */
function longestCommon(before: readonly string[], after: readonly string[]): number {
  let previous = new Int32Array(after.length + 1);
  let current = new Int32Array(after.length + 1);

  for (let i = 1; i <= before.length; i += 1) {
    for (let j = 1; j <= after.length; j += 1) {
      current[j] =
        before[i - 1] === after[j - 1]
          ? previous[j - 1] + 1
          : Math.max(previous[j], current[j - 1]);
    }

    [previous, current] = [current, previous];
  }

  return previous[after.length];
}

/** How many tokens the engine managed to pair up. */
function paired(before: readonly string[], after: readonly string[]): number {
  return diffSequence(before, after)
    .filter((edit) => edit.kind === 'equal')
    .reduce((total, edit) => total + (edit.beforeEnd - edit.beforeStart), 0);
}

/** A sequence drawn from a small alphabet, so that runs repeat the way real text does. */
function sequence(random: () => number, length: number, alphabet: string): string[] {
  return Array.from({ length }, () => alphabet[Math.floor(random() * alphabet.length)]);
}

/** The same sequences on every run, so a failure can be looked at again. */
function seeded(seed: number): () => number {
  let state = seed;

  return () => {
    state = (state * 1103515245 + 12345) % 2147483648;

    return state / 2147483648;
  };
}

describe('the engine', () => {
  it('pairs up every token two identical sequences have', () => {
    const tokens = 'a b c d e'.split(' ');

    expect(paired(tokens, tokens)).toBe(5);
  });

  it('pairs up nothing between two sequences with nothing in common', () => {
    expect(paired(['a', 'b'], ['c', 'd'])).toBe(0);
  });

  it('finds as much in common as there is to find, over a thousand random pairs', () => {
    const random = seeded(20260905);

    for (let round = 0; round < 1000; round += 1) {
      const alphabet = round % 2 === 0 ? 'ab' : 'abcdef';
      const before = sequence(random, Math.floor(random() * 40), alphabet);
      const after = sequence(random, Math.floor(random() * 40), alphabet);

      expect({ round, paired: paired(before, after) }).toEqual({
        round,
        paired: longestCommon(before, after)
      });
    }
  });

  it('finds as much in common as there is to find in a long pair', () => {
    const random = seeded(7);
    const before = sequence(random, 2000, 'abcdefghij');
    const after = before.filter(() => random() > 0.1);

    expect(paired(before, after)).toBe(longestCommon(before, after));
  });

  it('leaves the two sequences reconstructable from the edits it returns', () => {
    const random = seeded(99);

    for (let round = 0; round < 200; round += 1) {
      const before = sequence(random, Math.floor(random() * 30), 'abc');
      const after = sequence(random, Math.floor(random() * 30), 'abc');
      const edits = diffSequence(before, after);

      const rebuiltBefore: string[] = [];
      const rebuiltAfter: string[] = [];

      for (const edit of edits) {
        rebuiltBefore.push(...before.slice(edit.beforeStart, edit.beforeEnd));
        rebuiltAfter.push(...after.slice(edit.afterStart, edit.afterEnd));
      }

      expect(rebuiltBefore).toEqual(before);
      expect(rebuiltAfter).toEqual(after);
    }
  });
});

File: packages/flutter/test/diff/myers_test.dart
import 'dart:math' as math;

import 'package:diffine/diffine.dart';
import 'package:diffine/src/internal/diff/myers.dart';
import 'package:flutter_test/flutter_test.dart';

/// How many tokens the two sequences have in common, as the search found them.
int paired(List<DiffMatch> matches) {
  return matches.fold(0, (int total, DiffMatch match) => total + match.length);
}

/// The longest run of tokens both sequences hold in order, worked out the slow
/// way — which is the answer the engine has to reach by the fast one.
int longestCommon(List<String> before, List<String> after) {
  final List<List<int>> table = List<List<int>>.generate(
    before.length + 1,
    (int _) => List<int>.filled(after.length + 1, 0),
  );

  for (int i = 1; i <= before.length; i += 1) {
    for (int j = 1; j <= after.length; j += 1) {
      table[i][j] = before[i - 1] == after[j - 1]
          ? table[i - 1][j - 1] + 1
          : math.max(table[i - 1][j], table[i][j - 1]);
    }
  }

  return table[before.length][after.length];
}

void main() {
  group('matchSequences', () {
    test('pairs up every token two identical sequences have', () {
      final List<String> tokens = 'a b c d e'.split(' ');

      expect(paired(matchSequences(tokens, tokens, 5000).matches), 5);
    });

    test('pairs up nothing between two sequences with nothing in common', () {
      expect(paired(matchSequences('a b c'.split(' '), 'x y z'.split(' '), 5000).matches), 0);
    });

    test('finds as much in common as there is to find, over a thousand random pairs', () {
      final math.Random random = math.Random(20260910);

      for (int round = 0; round < 1000; round += 1) {
        final List<String> before = List<String>.generate(
          random.nextInt(9),
          (int _) => String.fromCharCode(97 + random.nextInt(4)),
        );
        final List<String> after = List<String>.generate(
          random.nextInt(9),
          (int _) => String.fromCharCode(97 + random.nextInt(4)),
        );

        expect(
          paired(matchSequences(before, after, 5000).matches),
          longestCommon(before, after),
          reason: '$before vs $after',
        );
      }
    });

    test('finds as much in common as there is to find in a long pair', () {
      final math.Random random = math.Random(7);
      final List<String> before = List<String>.generate(
        400,
        (int index) => 'line ${random.nextInt(120)}',
      );
      final List<String> after = List<String>.from(before)
        ..removeRange(30, 60)
        ..insertAll(120, <String>['one', 'two', 'three']);

      expect(paired(matchSequences(before, after, 5000).matches), longestCommon(before, after));
    });

    test('returns its runs in order and without overlaps', () {
      final List<DiffMatch> matches = matchSequences(
        'a b c d e f'.split(' '),
        'a x c d y f'.split(' '),
        5000,
      ).matches;

      int before = 0;
      int after = 0;

      for (final DiffMatch match in matches) {
        expect(match.beforeStart, greaterThanOrEqualTo(before));
        expect(match.afterStart, greaterThanOrEqualTo(after));
        before = match.beforeStart + match.length;
        after = match.afterStart + match.length;
      }
    });

    test('leaves the two sequences reconstructable from the edits it returns', () {
      final List<String> before = 'the quick brown fox jumps over'.split(' ');
      final List<String> after = 'the slow brown fox leaps over again'.split(' ');
      final List<String> rebuiltBefore = <String>[];
      final List<String> rebuiltAfter = <String>[];

      for (final DiffEdit edit in diffSequence(before, after)) {
        rebuiltBefore.addAll(before.sublist(edit.beforeStart, edit.beforeEnd));
        rebuiltAfter.addAll(after.sublist(edit.afterStart, edit.afterEnd));
      }

      expect(rebuiltBefore, before);
      expect(rebuiltAfter, after);
    });
  });
}
